### ISSUE TITLE
Resilience for workflows that end in `to_<storage>`

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#8182]
# Use case
An embarrassingly parallel workflow, where the number of independent pipelines >> the number of workers, ends with a call to `to_zarr` or some other of form of permanent storage. At some point during the computation, a single worker - let's say one out of 16-  crashes unexpectedly.

### Expected behaviour
You lose the pipelines that were halfway through on that worker. They are rescheduled automatically.

### Actual behaviour
You lose 1/16th of *all your work so far*. The scheduler wastes time recomputing data that has already been safely saved to your permanent storage.

### What happened
`to_<storage>` calls generate one task per chunk/partitiion, each emitting `None` as output, which is stored on the workers until the whole computation is complete. The scheduler doesn't do anything special about them - which means that if the worker holding them crashes the scheduler will recompute the whole pipeline leading to them.

In pseudocode:
```python
a = dask.array.zeros(...)
a = a.map_blocks(expensive_processing)
z = a.map_blocks(to_storage, fname="s3://somewhere")
z = z.persist()
client.wait(z)
client.gather(z)
```

# Proposed solution 1: AMM AutoReplicate
Implement a new Active Memory Manager policy, `AutoReplicate`, which automatically creates e.g. 4 replicas for tasks whose sizeof is < 1 kiB. Exclude tasks that have between 1 and 4 dependants.

- XREF https://github.com/dask/distributed/issues/6578

# Proposed solution 2: rewrite graph
If the sizeof of the output of a task is e.g. < 500 bytes AND its pickled size is also < 500 bytes, then the `TaskFinishedMsg` will contain the actual pickled value of the task. The scheduler will then immediately
- change the `run_spec` to `lambda: <output>`
- release all the task's non-shared dependencies

If the worker later dies, the scheduler will just reschedule a trivial `lambda: <output>`
You would see the total number of tasks on the scheduler decrease progressively as the computation progresses.
It also means that large amounts of workers could die at the same time without losing work.

CC @fjetter @hendrikmakait

**Comment by mrocklin** [comment]
Hrm, I hadn't ever thought of that.  Yes, I can see how that would suck.

**Comment by d-v-b** [comment]
Would it make anything easier if the storage routines returned something more informative than `None`? At least in the case of writing arrays, a return value conveying the region / chunks written would be a natural way of tracking progress.

**Comment by fjetter** [comment]
Yes, I am aware of this problem. In the past I once wrote some logic that would check (e.g. lookup in a database) whether a certain task already ran before it ran and would execute immediately if it was the case. This kind of logic would be an _easy_ SchedulerPlugin

> If the sizeof of the output of a task is e.g. < 500 bytes AND its pickled size is also < 500 bytes, then the TaskFinishedMsg will contain the actual pickled value of the task. The scheduler will then immediately

I think this idea has potential beyond data sink tasks. However, I wouldn't modify task states or release tasks or smth like this. That smells like trouble (state machine stuff).
I would likely suggest to use the _scheduler_ as a storage medium for these kinds of tasks. That would be a bit of a paradigm shift but would simultaneously also allow checkpointing for all other kinds of workflows and I believe the logic could be easily embedded into a nice extension to not mess with scheduler internals too much (we'd need to change some logic around `who_has` and such things but that'd be it)

**Comment by crusaderky** [comment]
> I would likely suggest to use the _scheduler_ as a storage medium for these kinds of tasks. That would be a bit of a paradigm shift but would simultaneously also allow checkpointing for all other kinds of workflows and I believe the logic could be easily embedded into a nice extension to not mess with scheduler internals too much (we'd need to change some logic around `who_has` and such things but that'd be it)

This idea was actually in the first draft of my post.
I abandoned it as it would introduce a wealth of new concepts (is the scheduler a fake worker now? or do we have a task in memory with empty who_has?) that would require a lot of legwork to implement. Not impossible, mind you, and definitely cleaner than pushing a dummy task to a worker, but definitely more _complex_. After all, we already have the concept of non-executable keys in the graph whose final value is stored on the scheduler (e.g. `da.from_numpy`); the only difference (which shouldn't really matter) is that they're not keys that clients will typically want to retrieve; so the effort should be minimal.

**Comment by crusaderky** [comment]
The idea of AMM AutoReplicate is not without merit and could actually cohexist with the graph rewrite. Consider:
```python
a = da.random.random((2**20, 2**20))
b = a.map_blocks(expensive)
c = b.sum(axis=0)
```
In this case, the partials at the end of the embarassingly parallell section are 8 MiB each, so it's out of question to store them on the scheduler. However, an AMM policy could say "if the fraction of memory usage / cumulative time needed to compute this key and all its dependencies is below a certain threshold, replicate it for safekeeping". 

In other words, if the risk of losing a key outweighs the cost of keeping extra copies laying around, spend some RAM to build redundancy.
